jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java

```diff
@@ -580,34 +583,27 @@ public void onTimeoutExpired()
                     continue;
                 if (expiresAt <= now)
                     request.abort(new TimeoutException("Total timeout " + request.getTimeout() + " ms elapsed"));
-                else if (expiresAt < nextExpiresAt)
-                    nextExpiresAt = expiresAt;
+                else if (expiresAt < earliest)
+                    earliest = expiresAt;
             }
 
-            if (nextExpiresAt < Long.MAX_VALUE && client.isRunning())
-                schedule(nextExpiresAt);
+            if (earliest < Long.MAX_VALUE && client.isRunning())
+                schedule(earliest);
         }
 
         private void schedule(long expiresAt)
         {
-            // Schedule a timeout for the soonest any known exchange can expire.
-            // If subsequently that exchange is removed from the queue, the
-            // timeout is not cancelled, instead the entire queue is swept
-            // for expired exchanges and a new timeout is set.
-            long timeoutAt = nextTimeout.getAndUpdate(e -> Math.min(e, expiresAt));
-            if (timeoutAt != expiresAt)
+            // Schedule a timeout for the earliest exchange that may expire.
+            // When the timeout expires, scan the exchange queue for the next
+            // earliest exchange that may expire, and reschedule a new timeout.
+            long earliest = earliestTimeout.getAndUpdate(t -> Math.min(t, expiresAt));
+            if (expiresAt < earliest)
```

Comment:
Is this right?   If expiresAt is after earliestTimeout, then earliest == earliestTimeout, which is < expiresAt, so do nothing... correct.
But if expiresAt is before earliestTimeout, then earliest == expiresAt, so the < also fails and we also do nothing - wrong?

Surely we need to do a schedule IFF earliest==expiresAt ?

```suggestion
            if (expiresAt == earliest)
```

Reply:
@gregw the update is a get-then-update, so `earliest` is the _old_ minimum. We don't know the new minimum, could be either `earliest` or `expireAt`.

The less-than check determines what the new minimum is; if it's `expireAt` then we need to schedule.

Reply:
Ah, then `earliest` is the wrong name.  Should be `lastEarliest`

Reply:
or `expiringAt`

Reply:
`prevEarliest`